### PR TITLE
Dedicated thread Scheduler

### PR DESCRIPTION
**Pull request by rogeralsing**
This is a PoC to see what happens when the scheduler runs on a dedicated thread.
When running with the TaskBasedScheduler (the default) and the thread pool gets flooded, Akka.Remote can no longer push Heartbeats through and endpoints gets disassociated.

When using this scheduler, the disassociation no longer occur.

I have tested this by letting the chat application spin up 600 blocking actors on a special command.

scheduler

That turned out to delay Heartbeats by almost 40 seconds.

When running the new scheduler, everything works fine in my test.

### CODE REVIEW

**Comment by rogeralsing** [comment]
cc @Aaronontheweb @annymsMthd

File: src/core/Akka/Actor/Scheduler/DedicatedThreadScheduler.cs [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+            var expected = DateTime.UtcNow + initialDelay;
+            Action executeAction = () =>
+            {
+                if (token.IsCancellationRequested)
+                    return;
+
+                try
+                {
+                    var actual = DateTime.UtcNow;
+                    Console.WriteLine("InternalScheduleTellOnce diff {0}", (actual - expected).TotalMilliseconds);
```
Comment (rogeralsing) [inline src/core/Akka/Actor/Scheduler/DedicatedThreadScheduler.cs]: debugging info just to see how this performs..

**Comment by annymsMthd** [comment]
This does the trick.

File: src/core/Akka/Actor/Scheduler/DedicatedThreadScheduler.cs [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+        protected override void InternalScheduleRepeatedly(TimeSpan initialDelay, TimeSpan interval, Action action, ICancelable cancelable)
+        {
+            var cancellationToken = cancelable == null ? CancellationToken.None : cancelable.Token;
+            InternalScheduleRepeatedly(initialDelay, interval, action, cancellationToken);
+        }
+
+
+        private void InternalScheduleOnce(TimeSpan initialDelay, Action action, CancellationToken token)
+        {
+            var expected = DateTime.UtcNow + initialDelay;
```
Comment (Aaronontheweb) [inline src/core/Akka/Actor/Scheduler/DedicatedThreadScheduler.cs]: can we modify this to start using the monotonic clock per #846?

**Comment by rogeralsing** [comment]
I've updated the code to use monotonic clock, and removed the debug writelines.

We need some sort of mechanism to kill the thread when the system exits.
Do we have one that would be appropriate here?
`system.what?`

**Comment by rogeralsing** [comment]
We should probably pass a Config to the pool to, and read the precision value from there

**Comment by Aaronontheweb** [comment]
@rogeralsing you can use the termination hooks for that

**Comment by rogeralsing** [comment]
`system.TerminationTask` or something else?

**Comment by rogeralsing** [comment]
Even if this is largely untested, it's still better than what we have right now.
Do we pull this in or should someone port the JVM scheduler instead?

File: src/core/Akka/Actor/Internals/ActorSystemImpl.cs [hunk 3]
```diff
@@ -51,11 +51,11 @@ public ActorSystemImpl(string name, Config config)
             if(config == null)
                 throw new ArgumentNullException("config");
 
-            _name = name;
-            ConfigureScheduler();
+            _name = name;            
             ConfigureSettings(config);
             ConfigureEventStream();
             ConfigureProvider();
+            ConfigureScheduler();
```
Comment (rogeralsing) [inline src/core/Akka/Actor/Internals/ActorSystemImpl.cs]: provider and settings needs to be in place before we can config the scheduler

**Comment by annymsMthd** [comment]
I'm all for this. Tests are good. Did any check the multinode tests?

**Comment by rogeralsing** [comment]
Nope, my VS + xUnit is giving me a really hard time ATM. cannot compile any tests myself right now

**Comment by Aaronontheweb** [comment]
I'm all for bringing this in. Death to unnecessary disassociations!

**Comment by Aaronontheweb** [comment]
I'll run the MultiNode tests now

**Pull request merged**